examples/matching-engine/tests/transaction.rs

```diff
@@ -0,0 +1,263 @@
+        let order = Order::Insert {
+            owner: owner_a,
+            amount: Amount::from_tokens(3),
+            nature: OrderNature::Bid,
+            price,
+        };
+        let operation = Operation::ExecuteOrder { order };
+        let order_messages = fungible_chain_a
+            .add_block(|block| {
+                block.with_operation(matching_id, operation);
+            })
+            .await;
+
+        assert_eq!(order_messages.len(), 3);
+        orders_bids.extend(order_messages);
+    }
+
+    matching_chain
+        .add_block(|block| {
+            block.with_incoming_messages(orders_bids);
+        })
+        .await;
+
+    fungible_chain_a.handle_received_messages().await;
+    fungible_chain_b.handle_received_messages().await;
+    matching_chain.handle_received_messages().await;
+
+    // Checking the values for token_a
+    for (owner, amount) in [
+        (admin_account, None),
+        (owner_a, Some(Amount::from_tokens(1))),
+        (owner_b, None),
+    ] {
+        let value = FungibleTokenAbi::query_account(token_id_a, &fungible_chain_a, owner).await;
+        assert_eq!(value, amount);
+    }
+    for (owner, amount) in [(admin_account, None), (owner_a, None), (owner_b, None)] {
+        assert_eq!(
+            FungibleTokenAbi::query_account(token_id_a, &fungible_chain_b, owner).await,
+            amount
+        );
+    }
+    for (owner, amount) in [
+        (admin_account, None),
+        (owner_a, Some(Amount::ZERO)),
+        (owner_b, None),
+    ] {
+        assert_eq!(
+            FungibleTokenAbi::query_account(token_id_a, &matching_chain, owner).await,
+            amount
+        );
+    }
+
+    let mut orders_asks = Vec::new();
+    for price in [4, 2] {
+        let price = Price { price };
+        let order = Order::Insert {
+            owner: owner_b,
+            amount: Amount::from_tokens(4),
+            nature: OrderNature::Ask,
+            price,
+        };
+        let operation = Operation::ExecuteOrder { order };
+        let order_messages = fungible_chain_b
+            .add_block(|block| {
+                block.with_operation(matching_id, operation);
+            })
+            .await;
+
+        assert_eq!(order_messages.len(), 3);
+        orders_asks.extend(order_messages);
+    }
+
+    matching_chain
+        .add_block(|block| {
+            block.with_incoming_messages(orders_asks);
+        })
+        .await;
+
+    fungible_chain_a.handle_received_messages().await;
+    fungible_chain_b.handle_received_messages().await;
+    matching_chain.handle_received_messages().await;
+
+    // Now querying the matching
+    let order_ids_a = get_orders(matching_id, &matching_chain, owner_a)
+        .await
+        .expect("order_ids_a");
+    assert_eq!(order_ids_a.len(), 1);
+    let order_ids_b = get_orders(matching_id, &matching_chain, owner_b)
+        .await
+        .expect("order_ids_b");
+    assert_eq!(order_ids_b.len(), 2);
```

Comment:
So if I followed this correctly:

* Owner A offered 6 A for 3 B, and another 3 A for 3 B?
* Owner B offered 4 B for 8 A, and another 4 B for 16 A?
* So B bought 6 A for 3 B from A, completely filling one of A's orders, but only partly one of B's orders?
* That's why A has one order left, and B two?

Would be helpful to explain that in the comments.

Reply:
Exactly. Added the information in the text.